### PR TITLE
[#471] Select parent institution modal

### PR DESCRIPTION
**Pull request by danielkryska**
**IMPORTANT!!! Can be run by going to `settings` page and clicking first `Dashboard` tab on sidebar**

[Preview](https://projects.invisionapp.com/share/S9RJNE8WXUB#/screens/416623692)

* Modal with institution selection
* Search institution
* Show filtered institutions

Closes: #471

### CODE REVIEW

File: s4e-web/src/app/state/form/form.model.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ import {SaveConfigForm} from '../../views/map-view/zk/save-config-modal/save-con
 import {ReportForm} from '../../views/map-view/zk/report-modal/report-modal.model';
 import {SentinelSearchForm} from '../../views/map-view/state/sentinel-search/sentinel-search.model';
 import {ShareConfigurationForm} from '../../views/map-view/zk/configuration/state/configuration.model';
+import { ParentInstitutionForm } from 'src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.model';
```
Comment (michal-szostak) [inline s4e-web/src/app/state/form/form.model.ts:10]: The style is to not have spaces after curly braces eg. `{ParentInstitutionForm}`

File: s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+@Component({
+  selector: 's4e-parent-institution-modal',
+  templateUrl: './parent-institution-modal.component.html',
+  styleUrls: ['./parent-institution-modal.component.scss']
+})
+export class ParentInstitutionModalComponent extends ModalComponent<Institution> implements OnInit, OnDestroy {
+  public institutions: Institution[] = [];
+  public searchedInstitutions: Institution[] = [];
+  public selectedInstitution: Institution = null;
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts:26]: Generally if we accept that variable can be null then it's better to use type like this: `Institution|null`

File: s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    private _institutionService: InstitutionService,
+    @Inject(MODAL_DEF) modal: ParentInstitutionModal
+  ) {
+    super(_modalService, PARENT_INSTITUTION_MODAL_ID);
+
+    assertModalType(isParentInstitutionModal, modal);
+  }
+
+  ngOnInit() {
+    this._institutionService.get();
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts:42]: Calling `get` or other service methods which load data is better done after all subscribes are set up - so you should move this line to the end of `ngOnInit()` method

File: s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    this.institutionsSearch.valueChanges
+      .pipe(untilDestroyed(this))
+      .subscribe(() => {
+        this.searchedInstitutions = this.institutions
+          .filter(institution => this._isSearchedInstitution(institution));
+      });
+  }
+
+  accept() {
+    this.dismiss(this.selectedInstitution);
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts:59]: Is returning null an option? If not then you should check whether `selectedInstitution != null`

File: s4e-web/src/app/views/settings/settings.component.ts [hunk 5]
```diff
@@ -1,19 +1,51 @@
...
+      hasReturnValue: true
+    });
+
+    this._modalQuery.modalClosed$(PARENT_INSTITUTION_MODAL_ID)
+      .pipe(
+        untilDestroyed(this),
+        map(modal => isParentInstitutionModal(modal) ? modal.returnValue : null)
+      )
+      .toPromise()
+      .then((institution: Institution) => console.log(institution));
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/settings.component.ts:47]: This is unnecessary - you can just use `.subscribe()`.

File: s4e-web/src/app/views/settings/settings.component.ts [hunk 6]
```diff
@@ -1,19 +1,51 @@
...
+  openParentInstitutionModal() {
+    this._modalService.show<ParentInstitutionModal>({
+      id: PARENT_INSTITUTION_MODAL_ID,
+      size: 'lg',
+      hasReturnValue: true
+    });
+
+    this._modalQuery.modalClosed$(PARENT_INSTITUTION_MODAL_ID)
+      .pipe(
+        untilDestroyed(this),
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/settings.component.ts:43]: this is unnecessary, `.modalClosed$` returns observable which completes

File: s4e-web/src/app/views/settings/settings.component.html [hunk 7]
```diff
@@ -1,6 +1,10 @@
 <div class="container">
   <nav class="navigation">
     <ul class="navigation__list">
+      <li *ngIf="showInstitutions$ | async">
+        <a href="javascript:void(0)" (click)="openParentInstitutionModal()">Dashboard</a>
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/settings.component.html:5]: I know its temporary, but maybe change the name of the link to `show institution modal` or something

File: s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        this.institutions = institutions;
+        this.searchedInstitutions = this.institutions;
+      });
+
+    this.institutionsSearch.valueChanges
+      .pipe(untilDestroyed(this))
+      .subscribe(() => {
+        this.searchedInstitutions = this.institutions
+          .filter(institution => this._isSearchedInstitution(institution));
+      });
```
Comment (michal-szostak) [inline s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts:55]: I would suggest connecting these two observables into one. The race condition can in theory occur here.

The workflow should be that you should `combineLatest` from search query and institutions.

File: s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts (hunk 3 above)
Comment (danielkryska) [reply s4e-web/src/app/views/settings/manage-institutions/parent-institution-modal/parent-institution-modal.component.ts:42]: Right

File: s4e-web/src/app/state/form/form.model.ts (hunk 1 above)
Comment (danielkryska) [reply s4e-web/src/app/state/form/form.model.ts:10]: It's automated by VS to add it this way. As it is inconsistent trough the whole application code, maybe instead repairing it in here, we should use some pre-commit `linter`.

**Pull request merged**